### ISSUE TITLE
KeyError: ber_class 2, ber_type 3

### ISSUE THREAD

**Issue by TWAC** [cannatag/ldap3#133]
I have an issue when querying an Active Directory server.
Older ldap3 versions works, bisection shows the issue appearing between ldap3 0.9.8.8 and 0.9.9.

When the query hits an user located in a another domain, ldap3 hits a KeyError exception.
```text
  File "C:\Python34\lib\site-packages\ldap3\core\connection.py", line 614, in search
    response = self.post_send_search(self.send('searchRequest', request, controls))
  File "C:\Python34\lib\site-packages\ldap3\strategy\sync.py", line 142, in post_send_search
    responses, result = self.get_response(message_id)
  File "C:\Python34\lib\site-packages\ldap3\strategy\base.py", line 301, in get_response
    responses = self._get_response(message_id)
  File "C:\Python34\lib\site-packages\ldap3\strategy\sync.py", line 167, in _get_response
    ldap_resp = decode_message_fast(response)
  File "C:\Python34\lib\site-packages\ldap3\utils\asn1.py", line 72, in decode_message_fast
    decoded = decode_sequence(message, ber_value_offset, ber_len + ber_value_offset, LDAP_MESSAGE_CONTEXT)
  File "C:\Python34\lib\site-packages\ldap3\utils\asn1.py", line 92, in decode_sequence
    value = ber_decoder(message, start, start + ber_len, context_decoders)  # call value decode function
  File "C:\Python34\lib\site-packages\ldap3\utils\asn1.py", line 94, in decode_sequence
    value = context_decoders[ber_type](message, start, start + ber_len)  # call value decode function for context class
KeyError: 3
```
I could make it work again by simply decoding the message using decode_octet_string.
Anonymized, the message looks like this:
```python
server == "local.domain.name"

ber_class == 2
ber_type == 3
value == b'\x04oldaps://other.domain.name/OU=Users,DC=other,DC=domain,DC=name'
```

```patch
--- a/ldap3/utils/asn1.py
+++ b/ldap3/utils/asn1.py
@@ -90,8 +90,10 @@ def decode_sequence(message, start, stop, context_decoders=None):
         start += ber_value_offset
         if ber_decoder:
             value = ber_decoder(message, start, start + ber_len, context_decoders)  # call value decode function
-        else:
+        elif ber_type in context_decoders:
             value = context_decoders[ber_type](message, start, start + ber_len)  # call value decode function for context class
+        else:
+            value = decode_octet_string(message, start, start + ber_len)
         decoded.append((ber_class, ber_constructed, ber_type, value))
         start += ber_len
 
```

**Comment by cannatag** [comment]
Hi, the main change between those release is a "fast" custom ber decoder. You can use the decoder supplied by pyasn1 with ``fast_decoder=False`` while defining the connection object. I will check your fix and include it in the next release of ldap3.

Giovanni

**Comment by TWAC** [comment]
Thanks, using `fast_decoder=False` works fine. The new ldap3 version is quite a bit faster anyway.

**Comment by cannatag** [comment]
Hi, I've fix this issue in a slight different way from what you proposed.
There was a problem with referrals not being properly decoded, Now they are working as expected. You can try the "auto referral" feature of ldap3 specifying ``allowed_referral_hosts=[('*', True)]`` in the server object and ``auto_referrals=True`` in the connection object.

Thse spped of the fast decoder is nof affected by this change. You can find the fix in dev. Let me know if this works for you and I will release it soon. 

Thanks,
Giovanni

**Comment by TWAC** [comment]
Hello, thanks again, the dev branch works fine without `fast_decoder=False`, and is very fast.

I have tried with and without `allowed_referral_hosts=[('*', True)]` and `auto_referrals=True`. It makes no difference to the search, still returns False. I suspect something in my configuration is not set up correctly to allow referrals.

**Comment by cannatag** [comment]
Hi TWAC,
in my tests the referrals work correctly. Coutd you try to get a log with the NETWORK level? (if your query data are not sensitive) It should made clear where the referrals are pointing to. Also, can you give me back the value of ``connection.result``with ``auto_referrals=False`` in the connection object?

Thanks,
Giovanni

**Comment by cannatag** [comment]
Released in 1.0.1
